Problem serializing to JSON when upgrading from 2.8.2 to 2.8.3

### Play Version

2.8.3 

### API

Java API

### Operating System

Linux NBSL1380 5.4.0-52-generic #57-Ubuntu SMP Thu Oct 15 10:57:00 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux

### JDK

openjdk version "11.0.8" 2020-07-14
OpenJDK Runtime Environment AdoptOpenJDK (build 11.0.8+10)
OpenJDK 64-Bit Server VM AdoptOpenJDK (build 11.0.8+10, mixed mode)


### Library Dependencies

<!-- If this is an issue that involves integration with another system, include the exact version and OS of the other system, including any intermediate drivers or APIs i.e. if you connect to a PostgreSQL database, include both the version / OS of PostgreSQL and the JDBC driver version used to connect to the database. -->

### Expected Behavior

`play.libs.Json.toJson` is able to serialise a Jackson serializable object as it was in 2.8.2

### Actual Behavior

Our tests now fail with stack trace provided below: 
```
      java.lang.RuntimeException: java.lang.RuntimeException: java.lang.RuntimeException: We should have been reading an object, something got wrong
        at controllers.DecisionStrategyController.lambda$getDecisionStrategy$0(DecisionStrategyController.java:50) Test / executeTests 19s
	at java.base/java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:642)
	at java.base/java.util.concurrent.CompletableFuture$Completion.run(CompletableFuture.java:478)
	at play.core.j.HttpExecutionContext.$anonfun$execute$1(HttpExecutionContext.scala:64)
	at akka.dispatch.TaskInvocation.run(AbstractDispatcher.scala:48)
	at akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask.exec(ForkJoinExecutorConfigurator.scala:48)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:183)
	at ✽.I should have a strategy with proxy settings attached(file:///home/tom/deko-docker/src/decisioning/test/cucumber/features/ProxySettings.feature:13)
Caused by: java.lang.RuntimeException: java.lang.RuntimeException: We should have been reading an object, something got wrong
	at play.libs.Json.toJson(Json.java:94)
	at controllers.DecisionStrategyController.lambda$getDecisionStrategy$0(DecisionStrategyController.java:48)
	at java.base/java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:642)
	at java.base/java.util.concurrent.CompletableFuture$Completion.run(CompletableFuture.java:478)
	at play.core.j.HttpExecutionContext.$anonfun$execute$1(HttpExecutionContext.scala:64)
	at akka.dispatch.TaskInvocation.run(AbstractDispatcher.scala:48)
	at akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask.exec(ForkJoinExecutorConfigurator.scala:48)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:183)
Caused by: java.lang.RuntimeException: We should have been reading an object, something got wrong
	at play.utils.JsonNodeDeserializer.deserialize(JsonNodeDeserializer.scala:166)
	at play.utils.JsonNodeDeserializer.deserialize(JsonNodeDeserializer.scala:57)
	at play.utils.JsonNodeDeserializer.deserialize(JsonNodeDeserializer.scala:53)
	at com.fasterxml.jackson.databind.ObjectMapper._readValue(ObjectMapper.java:4497)
	at com.fasterxml.jackson.databind.ObjectMapper.readTree(ObjectMapper.java:2840)
	at com.fasterxml.jackson.databind.ObjectMapper.valueToTree(ObjectMapper.java:3229)
	at play.libs.Json.toJson(Json.java:92)
	... 11 more
```

Maybe a downstream issue with https://github.com/playframework/play-json?

### Reproducible Test Case

I'm unable to provide a reproducible test case at this time, I will try and spend some time later if possible and work out where the issue is.

--- a/core/play-integration-test/src/it/scala/play/it/libs/json/JavaJsonSpec.scala
+++ b/core/play-integration-test/src/it/scala/play/it/libs/json/JavaJsonSpec.scala
@@ -170,19 +170,16 @@ trait JavaJsonSpec extends Specification {
       "embedded/raw object" in new JsonScope {
 
         val expected =
-          """
-            |{
-            | "id": "1234",
+          """{"id":"abcd","contents":
             | "contents": { 
             |   "items": [
             |     {"id":"t-shirt", "count": 3},
             |     {"id":"mug",     "count": 5}
             |   ]
             |  }
-            |}
-            |""".stripMargin
+            |}""".stripMargin
 
-        val contents =
+        val raw =
           """
             | "contents": { 
             |   "items": [
@@ -190,9 +187,8 @@ trait JavaJsonSpec extends Specification {
             |     {"id":"mug",     "count": 5}
             |   ]
             |  }
-            |
             |""".stripMargin
-        val cart = new ShoppingCart("abcd", contents)
+        val cart = new ShoppingCart("abcd", raw)
         val json = Json.stringify(Json.toJson(cart))
 
         json must_== expected

--- a/core/play-integration-test/src/it/scala/play/it/libs/json/JavaJsonSpec.scala
+++ b/core/play-integration-test/src/it/scala/play/it/libs/json/JavaJsonSpec.scala
@@ -5,12 +5,12 @@
 package play.it.libs.json
 
 import java.io.ByteArrayInputStream
-import java.math.BigInteger
-import java.math.MathContext._
 import java.time.Instant
 import java.util.Optional
 import java.util.OptionalInt
 
+import com.fasterxml.jackson.annotation.JsonRawValue
+import com.fasterxml.jackson.annotation.JsonUnwrapped
 import com.fasterxml.jackson.databind.JsonNode
 import com.fasterxml.jackson.databind.ObjectMapper
 import com.fasterxml.jackson.databind.SerializationFeature
@@ -110,6 +110,23 @@ trait JavaJsonSpec extends Specification {
       "from InputStream" in new JsonScope {
         Json.parse(testJsonInputStream) must_== testJson
       }
+
+      "string into wrapped objects" in new JsonScope {
+        val json =
+          """
+            |{
+            | "code": "1234",
+            | "city": "NYC",
+            | "street": "Manhattan Av."
+            |}
+            |""".stripMargin
+
+        private val jsonNode: JsonNode = Json.parse(json)
+        private val customer: Customer = Json.fromJson(jsonNode, classOf[Customer])
+        customer.code must_== "1234"
+        customer.address must_== new Address("NYC", "Manhattan Av.")
+      }
+
     }
 
     "stringify" in {
@@ -141,6 +158,47 @@ trait JavaJsonSpec extends Specification {
         // The configured mapper should be able to handle Java Time fields
         Json.mapper().writeValueAsString(instant) must_== """"2015-03-04T02:24:21Z""""
       }
+
+      "unwrapped objects" in new JsonScope {
+        val c = new Customer("1234")
+        c.address = new Address("NYC", "Manhattan Av.")
+        val json = Json.stringify(Json.toJson(c))
+        json must_== """{"code":"1234","city":"NYC","street":"Manhattan Av."}"""
+
+      }
+
+      "embedded/raw object" in new JsonScope {
+
+        val expected =
+          """
+            |{
+            | "id": "1234",
+            | "contents": { 
+            |   "items": [
+            |     {"id":"t-shirt", "count": 3},
+            |     {"id":"mug",     "count": 5}
+            |   ]
+            |  }
+            |}
+            |""".stripMargin
+
+        val contents =
+          """
+            | "contents": { 
+            |   "items": [
+            |     {"id":"t-shirt", "count": 3},
+            |     {"id":"mug",     "count": 5}
+            |   ]
+            |  }
+            |
+            |""".stripMargin
+        val cart = new ShoppingCart("abcd", contents)
+        val json = Json.stringify(Json.toJson(cart))
+
+        json must_== expected
+
+      }
+
     }
 
     "when deserializing to a POJO" should {
@@ -213,5 +271,60 @@ trait JavaJsonSpec extends Specification {
         jsonNode.has("foo") must beTrue
       }
     }
+  }
+}
+
+/// Below this point there's a handful of classes useful to test certain Jackson edge cases. These classes
+/// are not wrapped on an object or use any scala-idiomatic feature on purpose. The reasons are (1) to ensure
+/// no Jackson modules for scala-specific code are interfering, and (2) to avoid hitting cases Jackson
+/// doesn't support such as inner classes being non-static or similar.
+class Address(var city: String, var street: String) {
+  override def toString             = s"Address(city=$city, street=$street)"
+  def canEqual(other: Any): Boolean = other.isInstanceOf[Address]
+  override def equals(other: Any): Boolean = other match {
+    case that: Address =>
+      (that.canEqual(this)) &&
+        city == that.city &&
+        street == that.street
+    case _ => false
+  }
+  override def hashCode(): Int = {
+    val state = Seq(city, street)
+    state.map(_.hashCode()).foldLeft(0)((a, b) => 31 * a + b)
+  }
+}
+class Customer(var code: String) {
+  @JsonUnwrapped var address: Address = null
+
+  override def toString             = s"Customer(code=$code, address=$address)"
+  def canEqual(other: Any): Boolean = other.isInstanceOf[Customer]
+  override def equals(other: Any): Boolean = other match {
+    case that: Customer =>
+      (that.canEqual(this)) &&
+        code == that.code &&
+        address == that.address
+    case _ => false
+  }
+  override def hashCode(): Int = {
+    val state = Seq(code, address)
+    state.map(_.hashCode()).foldLeft(0)((a, b) => 31 * a + b)
+  }
+}
+
+// @JsonRawValue is a serialization-only annotation so this model can only be used in serialization tests.
+// see https://github.com/FasterXML/jackson-annotations/wiki/Jackson-Annotations#serialization-details
+class ShoppingCart(var id: String, @JsonRawValue var contents: String) {
+  override def toString             = s"ShoppingCart(id=$id, contents=$contents)"
+  def canEqual(other: Any): Boolean = other.isInstanceOf[ShoppingCart]
+  override def equals(other: Any): Boolean = other match {
+    case that: ShoppingCart =>
+      (that.canEqual(this)) &&
+        id == that.id &&
+        contents == that.contents
+    case _ => false
+  }
+  override def hashCode(): Int = {
+    val state = Seq(id, contents)
+    state.map(_.hashCode()).foldLeft(0)((a, b) => 31 * a + b)
   }
 }

--- a/core/play/src/main/scala/play/utils/JsonNodeDeserializer.scala
+++ b/core/play/src/main/scala/play/utils/JsonNodeDeserializer.scala
@@ -111,6 +111,11 @@ private class JsonNodeDeserializer extends JsonDeserializer[JsonNode] {
 
   //====================================================================================
 
+  /* re-encoding of part of JsonNodeDeserializer (jackson-databind 2.10.5)
+   * https://github.com/FasterXML/jackson-databind/blob/jackson-databind-2.10.5/src/main/java/com/fasterxml/jackson/databind/deser/std/JsonNodeDeserializer.java#L602-L623
+   * TODO: remove when jackson-databind 2.12 is out
+   */
+
   private def fromEmbedded(p: JsonParser, ctxt: DeserializationContext): JsonNode = {
     val nodeFactory = ctxt.getNodeFactory
     import com.fasterxml.jackson.databind.JsonNode

--- a/core/play/src/main/scala/play/utils/JsonNodeDeserializer.scala
+++ b/core/play/src/main/scala/play/utils/JsonNodeDeserializer.scala
@@ -118,14 +118,21 @@ private class JsonNodeDeserializer extends JsonDeserializer[JsonNode] {
     val ob = p.getEmbeddedObject
 
     // should this occur?
-    if (ob == null) return nodeFactory.nullNode
-
-    val `type` = ob.getClass
-    if (`type` eq classOf[Array[Byte]]) return nodeFactory.binaryNode(ob.asInstanceOf[Array[Byte]]) // most common special case
-    if (ob.isInstanceOf[RawValue]) return nodeFactory.rawValueNode(ob.asInstanceOf[RawValue])
-    if (ob.isInstanceOf[JsonNode]) return ob.asInstanceOf[JsonNode]
-
-    nodeFactory.pojoNode(ob)
+    if (ob == null) {
+      nodeFactory.nullNode
+    } else {
+      val `type` = ob.getClass
+      if (`type` eq classOf[Array[Byte]]) {
+        // most common special case
+        nodeFactory.binaryNode(ob.asInstanceOf[Array[Byte]])
+      } else if (ob.isInstanceOf[RawValue]) {
+        nodeFactory.rawValueNode(ob.asInstanceOf[RawValue])
+      } else if (ob.isInstanceOf[JsonNode]) {
+        ob.asInstanceOf[JsonNode]
+      } else {
+        nodeFactory.pojoNode(ob)
+      }
+    }
   }
 
   //====================================================================================

--- a/core/play/src/main/scala/play/utils/JsonNodeDeserializer.scala
+++ b/core/play/src/main/scala/play/utils/JsonNodeDeserializer.scala
@@ -111,6 +111,25 @@ private class JsonNodeDeserializer extends JsonDeserializer[JsonNode] {
 
   //====================================================================================
 
+  private def fromEmbedded(p: JsonParser, ctxt: DeserializationContext): JsonNode = {
+    val nodeFactory = ctxt.getNodeFactory
+    import com.fasterxml.jackson.databind.JsonNode
+    import com.fasterxml.jackson.databind.util.RawValue
+    val ob = p.getEmbeddedObject
+
+    // should this occur?
+    if (ob == null) return nodeFactory.nullNode
+
+    val `type` = ob.getClass
+    if (`type` eq classOf[Array[Byte]]) return nodeFactory.binaryNode(ob.asInstanceOf[Array[Byte]]) // most common special case
+    if (ob.isInstanceOf[RawValue]) return nodeFactory.rawValueNode(ob.asInstanceOf[RawValue])
+    if (ob.isInstanceOf[JsonNode]) return ob.asInstanceOf[JsonNode]
+
+    nodeFactory.pojoNode(ob)
+  }
+
+  //====================================================================================
+
   @tailrec
   final def deserialize(
       jp: JsonParser,
@@ -162,8 +181,7 @@ private class JsonNodeDeserializer extends JsonDeserializer[JsonNode] {
       case JsonTokenId.ID_NOT_AVAILABLE =>
         throw new RuntimeException("Didn't receive a token when requesting one. See Jackson's JsonToken#NOT_AVAILABLE.")
 
-      case JsonTokenId.ID_EMBEDDED_OBJECT =>
-        throw new RuntimeException("Embedded JsonRawValue fields not supported.")
+      case JsonTokenId.ID_EMBEDDED_OBJECT => (Some(fromEmbedded(jp, ctxt)), parserContext)
     }
 
     // Read ahead

--- a/core/play/src/main/scala/play/utils/JsonNodeDeserializer.scala
+++ b/core/play/src/main/scala/play/utils/JsonNodeDeserializer.scala
@@ -160,10 +160,10 @@ private class JsonNodeDeserializer extends JsonDeserializer[JsonNode] {
         }
 
       case JsonTokenId.ID_NOT_AVAILABLE =>
-        throw new RuntimeException("We should have been reading an object, something got wrong")
+        throw new RuntimeException("Didn't receive a token when requesting one. See Jackson's JsonToken#NOT_AVAILABLE.")
 
       case JsonTokenId.ID_EMBEDDED_OBJECT =>
-        throw new RuntimeException("We should have been reading an object, something got wrong")
+        throw new RuntimeException("Embedded JsonRawValue fields not supported.")
     }
 
     // Read ahead

